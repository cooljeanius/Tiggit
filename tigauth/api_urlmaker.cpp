#include "api_urlmaker.hpp"
#include <assert.h>

using namespace TigAuth::API;

URLMaker::URLMaker(const std::string &_baseURL)
  : baseURL(_baseURL)
{
  // Make sure the base URL is slash terminated
  assert(baseURL.size() > 0);
  if(baseURL[baseURL.size() - 1] != '/')
    baseURL += "/";
}

std::string URLMaker::make(const std::string &base, bool q)
{
  std::string res = baseURL + base;
  if(key != "")
    {
      if(q) res += '?';
      else res += '&';
      q = false;
      res += "key=" + key;
    }
  if(wantlist != "")
    {
      if(q) res += '?';
      else res += '&';
      q = false;
      res += "want=" + wantlist;
    }
  if(userid != "")
    {
      if(q) res += '?';
      else res += '&';
      q = false;
      res += "asuser=" + userid;
    }

  return res;
}

std::string URLMaker::signal(const std::string &sign)
{ return make("signal_v1.php?act="+sign, false); }

std::string URLMaker::createBrowserLink()
{ return make("browser_v1.php"); }

std::string URLMaker::createJsonLink()
{ return make("json_v1.php"); }

std::string URLMaker::createSignOutLink()
{ return signal("signout"); }
