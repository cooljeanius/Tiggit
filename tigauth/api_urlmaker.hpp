#ifndef __TIGAUTH_API_URLMAKER_HPP_
#define __TIGAUTH_API_URLMAKER_HPP_

#include <string>

namespace TigAuth
{
  namespace API
  {
    struct URLMaker
    {
      // Input variables. Their meanings depend somewhat on what type
      // of URL is being produced.
      std::string key, wantlist, userid, baseURL;

      URLMaker(const std::string &_baseURL);

      /* Create a link that is meant to be opened in a web browser.
         The user has to perform some action manually on a web site.

         Parameters (either or both may be present):

         - key: if set, the website should ask the user to validate
           the given key for use with the API backend. The server may
           skip this step if the given key is already valid.

         - wantlist: if set, the website should redirect to a 'buy'
           page for all the items in the list. The list is a set of
           IDs separated by plus-signs.

         - userid: if set, sent to the website so it can double check
           whether the key's owner is the same the user logged in to
           the site.
       */
      std::string createBrowserLink();

      /* Create a request for more information from the API backend.
         This link is designed to be executed and decoded by
         software. The produced link will return a response coded in
         JSON containing the required information. Use
         api_response.hpp to decode it.

         Parameters:

         - key: must be set. Used to identify ourselves to the server

         - wantlist: may contain the ID of one item (not a list). If
           set, the API will return a one-time download link for the
           given item, or an error code if the item is not accessible.
           If NOT set, the API will instead return user information
           for the current user.
       */
      std::string createJsonLink();

      /* Create a one-way signout message to the server.

         Parameters:

         - key: the auth key to sign out. The key will be deleted from
           the server and will no longer work in future authentication
           requests.
       */
      std::string createSignOutLink();

    private:
      std::string make(const std::string &base, bool q=true);
      std::string signal(const std::string &sign);
    };
  }
}

#endif
