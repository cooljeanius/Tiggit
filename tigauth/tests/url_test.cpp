#include "api_urlmaker.hpp"
#include <iostream>

using namespace std;
using namespace TigAuth::API;

int main()
{
  cout << "Browser links:\n";
  {
    URLMaker url("test");
    url.key="KEY";
    cout << "  " << url.createBrowserLink() << endl;
    url.wantlist="hello+world";
    cout << "  " << url.createBrowserLink() << endl;
    url.key="";
    cout << "  " << url.createBrowserLink() << endl;
  }

  cout << "JSON links:\n";
  {
    URLMaker url("http://example.com/api/");
    url.key="KEY";
    cout << "  " << url.createJsonLink() << endl;
    url.wantlist="I+hope+you+are+fine";
    cout << "  " << url.createJsonLink() << endl;
  }

  cout << "Sign-out link:\n";
  {
    URLMaker url("http://someurl/with/long/path");
    url.key="KEY";
    cout << "  " << url.createSignOutLink() << endl;
  }

  return 0;
}
