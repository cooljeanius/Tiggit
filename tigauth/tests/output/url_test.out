Browser links:
  test/browser_v1.php?key=KEY
  test/browser_v1.php?key=KEY&want=hello+world
  test/browser_v1.php?want=hello+world
JSON links:
  http://example.com/api/json_v1.php?key=KEY
  http://example.com/api/json_v1.php?key=KEY&want=I+hope+you+are+fine
Sign-out link:
  http://someurl/with/long/path/signal_v1.php?act=signout&key=KEY
